File: PAT_Basic_Level_by_cpp/1114 全素日/1114 全素日.cpp
// 1114 全素日
//
// 实现原理：
// 给定一个按 yyyymmdd 格式给出的日期，判断它是否为"全素日"。
// 全素日是指：不仅日期本身是素数，而且它的任何以末尾数字结尾的子串都是素数。
// 即从完整 8 位数字串开始，按子串长度递减（8,7,...,1），每个子串都是原串的后缀。
//
// 关键步骤：
// 1. 对每个后缀子串（长度从 8 递减到 1），将其转换为整数。
// 2. 用试除法判断该整数是否为素数。
// 3. 每行输出子串和 Yes/No；若所有子串都是素数，最后输出 "All Prime!"。
//
// 复杂度分析：
// 时间复杂度：O(8 * sqrt(10^8))，子串最多 8 个，每个素数判断到 sqrt。
// 空间复杂度：O(1)，仅使用常数个变量。
#include <iostream>
#include <string>
#include <cmath>

using namespace std;

// 判断整数 x 是否为素数
bool isPrime(int x) {
    if (x < 2) return false;
    for (int i = 2; i * i <= x; ++i)
        if (x % i == 0) return false;
    return true;
}

int main() {
    string s;
    if (!(cin >> s)) return 0;

    bool all = true;
    for (size_t len = s.size(); len >= 1; --len) {
        string sub = s.substr(s.size() - len, len);  // 以末尾数字结尾的后缀子串
        int num = stoi(sub);                          // 转换为整数（忽略前导零）
        bool p = isPrime(num);
        if (!p) all = false;
        cout << sub << ' ' << (p ? "Yes" : "No") << endl;
    }
    if (all) cout << "All Prime!" << endl;
    return 0;
}

File: PAT_Basic_Level_by_java/1114 全素日/1114 全素日.java
/*
 * 1114 全素日
 * 
 * 题目描述：
 *     给定一个 8 位数字的日期字符串（如 "2023-01-01" 格式的 8 位数字，
 *     实际上输入是连续 8 位数字），从长度为 8 递减到长度为 1，
 *     分别取以最后一位结尾的后缀子串，判断每个子串对应的数字是否为素数。
 *     如果所有子串都是素数，则输出 "All Prime!"。
 * 
 * 实现原理：
 *     解题思路：
 *       1. 从长度 len = 8 递减到 1，取 date.substring(8 - len) 作为后缀子串。
 *       2. 将子串解析为 long 型整数，调用素数判断函数。
 *       3. 如果所有子串都是素数，最后输出 "All Prime!"。
 *     算法核心逻辑：
 *       - 素数判断：排除 1 及以下的非素数，2 是素数，偶数 >2 都不是素数，
 *         然后从 3 开始只检查奇数因子，上限为 sqrt(n)。
 *     时间复杂度：O(n * sqrt(M))，n=8 为固定长度，M 为最大数值。
 *     空间复杂度：O(1)。
 */
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        String date = br.readLine();
        
        boolean allPrime = true;

        // 从长度 8 递减到 1，依次取以最后一位结尾的后缀子串
        for (int len = date.length(); len >= 1; len--) {
            String subStr = date.substring(date.length() - len);
            long num = Long.parseLong(subStr);
            
            boolean isPrime = isPrime(num);
            
            System.out.println(subStr + " " + (isPrime ? "Yes" : "No"));

            if (!isPrime) {
                allPrime = false;
            }
        }

        if (allPrime) {
            System.out.println("All Prime!");

        }

        br.close();
    }
    
    /**
     * 判断一个数是否为素数
     * @param n 待判断的数
     * @return true 表示是素数，false 表示不是
     */
    private static boolean isPrime(long n) {
        if (n <= 1) {
            return false; // 1 及以下的数不是素数
        }
        if (n == 2) {
            return true; // 2 是素数
        }
        if (n % 2 == 0) {
            return false; // 偶数（除 2 外）不是素数
        }
        // 只检查奇数因子，上限为 sqrt(n)
        for (long i = 3; i * i <= n; i += 2) {
            if (n % i == 0) {
                return false;
            }
        }
        return true;
    }
}
